mantidproject/mantid
# Fix reflectivity calculation and error propagation in Q grouping in ILL reflectometry reduction

**Description of work.**

This PR reorganizes the `ReflectometryILL*` algorithms such that in the `SumInLambda` mode, the reflected data gets divided by the direct beam only after conversion to *Q* and grouping in *Q*. 

**Report to:** [nobody].

**To test:**

The script below exercises the workflow algorithms both in `SumInLambda` and `SumInQ` modes. The data is available in the unit test data directories. Check that the script runs and the two final workspaces, `reflectivity_sumInLambda` and `reflectivity_sumInQ` look about right; you should see something like this:

![reflectivity](https://user-images.githubusercontent.com/19147159/44922237-26cb3380-ad45-11e8-9afb-f8f0be9dc809.png)


```python
# Direct beam first
ReflectometryILLPreprocess('ILL/D17/317369.nxs',
    OutputWorkspace='direct',
    OutputBeamPositionWorkspace='direct_beampos')
ReflectometryILLSumForeground('direct',
    OutputWorkspace='direct_fgd',
    WavelengthRange=[2, 20.])
    
# SumInQ
ReflectometryILLPreprocess('ILL/D17/317370.nxs',
    OutputWorkspace='reflected',
    DirectBeamPositionWorkspace='direct_beampos')
ReflectometryILLSumForeground('reflected',
    OutputWorkspace='reflected_fgd_inQ',
    DirectForegroundWorkspace='direct_fgd',
    SummationType='SumInQ')
ReflectometryILLConvertToQ('reflected_fgd_inQ',
    OutputWorkspace='reflectivity_sumInQ',
    ReflectedBeamWorkspace='reflected',
    DirectBeamWorkspace='direct',
    GroupingQFraction=0.2)

# SumInLambda; same data as above
ReflectometryILLPreprocess('ILL/D17/317370.nxs',
    OutputWorkspace='reflected',
    DirectBeamPositionWorkspace='direct_beampos')
ReflectometryILLSumForeground('reflected',
    OutputWorkspace='reflected_fgd_inLambda',
    DirectForegroundWorkspace='direct_fgd')
ReflectometryILLConvertToQ('reflected_fgd_inLambda',
    OutputWorkspace='reflectivity_sumInLambda',
    ReflectedBeamWorkspace='reflected',
    DirectBeamWorkspace='direct',
    DirectForegroundWorkspace='direct_fgd',
    GroupingQFraction=0.2)
```

Fixes #22674.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
